File: product-of-array-except-self/solution238.go
// https://leetcode.com/problems/product-of-array-except-self/

package main

import (
	"fmt"
)

// T: O(n), S: O(n)
func productExceptSelf(nums []int) []int {
	n := len(nums)
	prefix := make([]int, n)
	left := 1
	for i := 0; i < n; i++ {
		prefix[i] = left
		left *= nums[i]
	}
	postfix := make([]int, n)
	right := 1
	for i := n - 1; i >= 0; i-- {
		postfix[i] = right
		right *= nums[i]
	}
	output := make([]int, n)
	for i := 0; i < n; i++ {
		output[i] = prefix[i] * postfix[i]
	}
	return output
}

// T: O(n), S: O(1)
func productExceptSelfOpt(nums []int) []int {
	n := len(nums)
	output := make([]int, n)
	left, right := 1, 1
	for i := 0; i < n; i++ {
		output[i] = left
		left *= nums[i]
	}
	for i := n - 1; i >= 0; i-- {
		output[i] *= right
		right *= nums[i]
	}
	return output
}

// T: O(n), S: O(1)
func productExceptSelfOpt2(nums []int) []int {
	n := len(nums)
	res := make([]int, n)
	for i := range res {
		res[i] = 1
	}

	leftProduct, rightProduct := 1, 1
	l, r := 0, n-1

	for l < n && r >= 0 {
		res[l] *= leftProduct
		res[r] *= rightProduct

		leftProduct *= nums[l]
		rightProduct *= nums[r]

		l++
		r--
	}

	return res
}

func main() {
	nums1 := []int{1, 2, 3, 4}
	nums2 := []int{-1, 1, 0, -3, 3}
	fmt.Println(productExceptSelf(nums1))     // [24, 12, 8, 6]
	fmt.Println(productExceptSelfOpt(nums1))  // [24, 12, 8, 6]
	fmt.Println(productExceptSelfOpt2(nums1)) // [24, 12, 8, 6]
	fmt.Println(productExceptSelf(nums2))     // [0, 0, 9, 0, 0]
	fmt.Println(productExceptSelfOpt(nums2))  // [0, 0, 9, 0, 0]
	fmt.Println(productExceptSelfOpt2(nums2)) // [0, 0, 9, 0, 0]
}

File: product-of-array-except-self/solution238.py
# https://leetcode.com/problems/product-of-array-except-self/

from typing import List


class Solution238:

    # T: O(n), S: O(n)
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        prefix = [1] * n
        left = 1
        for i in range(n):
            prefix[i] = left
            left *= nums[i]

        postfix = [1] * n
        right = 1
        for i in range(n - 1, -1, -1):
            postfix[i] = right
            right *= nums[i]

        output = [1] * n
        for i in range(n):
            output[i] = prefix[i] * postfix[i]

        return output

    # T: O(n), S: O(1)
    def productExceptSelfOpt(self, nums: List[int]) -> List[int]:
        n = len(nums)
        output = [1] * n
        left, right = 1, 1

        for i in range(n):
            output[i] = left
            left *= nums[i]

        for i in range(n - 1, -1, -1):
            output[i] *= right
            right *= nums[i]

        return output

    # T: O(n), S: O(1)
    def productExceptSelfOpt2(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n
        left_product, right_product = 1, 1
        l, r = 0, n - 1
        while l < n and r >= 0:
            res[l] *= left_product
            res[r] *= right_product
            left_product *= nums[l]
            right_product *= nums[r]
            l += 1
            r -= 1
        return res


def main() -> None:
    sol = Solution238()
    nums1 = [1, 2, 3, 4]
    nums2 = [-1, 1, 0, -3, 3]
    print(sol.productExceptSelf(nums1))    # [24, 12, 8, 6]
    print(sol.productExceptSelfOpt(nums1))  # [24, 12, 8, 6]
    print(sol.productExceptSelfOpt2(nums1))  # [24, 12, 8, 6]
    print(sol.productExceptSelf(nums2))    # [0, 0, 9, 0, 0]
    print(sol.productExceptSelfOpt(nums2))  # [0, 0, 9, 0, 0]
    print(sol.productExceptSelfOpt2(nums2))  # [0, 0, 9, 0, 0]


if __name__ == "__main__":
    main()
